### PR TITLE
Add missing __finalize__ calls in indexers/iterators

### PR DESCRIPTION
**Pull request by theOehrly**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This PR intends to add missing `__finalize__` calls to indexers and locators (this section of the docs: https://pandas.pydata.org/pandas-docs/stable/reference/frame.html#indexing-iteration). I went over all the methods listed in that section and in my opinion, everything should be covered after this PR.

This is generally in reference to #28283, although the methods covered here are not listed there. #19850 is also somewhat relevant.

The missing calls are only related to cases where a Series object was returned from indexing/iterating over a DataFrame. Though, I was somewhat generous with adding new test cases. Some newly added tests cases would have already passed before adding these calls, but I thought that it was still worth adding them.

### CODE REVIEW

**Review by phofl — COMMENTED** [review]
This broke existing tests

**Comment by theOehrly** [comment]
> This broke existing tests

Should be good now. I had to un-xfail some more tests that were previously marked as not implemented.
I just need to fix the small typing validation issue if there are no other things after review.

**Comment by theOehrly** [comment]
@phofl All tests are passing now. Can you review again, please?

**Comment by jreback** [comment]
looks pretty good @theOehrly if you can add a release note (I think we have a few already that are about finalize so if you can group with those).

**Comment by jreback** [comment]
cc @phofl if you can review when you have a chance

**Comment by theOehrly** [comment]
@jreback That one failure seems unrelated? At least I'm failing to make sense of it, and it did pass previously.
Correct me if I'm wrong, though.

File: doc/source/whatsnew/v1.4.2.rst [hunk 1]
```diff
@@ -23,6 +23,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
+- Fixed metadata propagation in :meth:`NDFrame.xs`, :meth:`DataFrame.iterrows`, :meth:`DataFrame.loc` and :meth:`DataFrame.iloc` (:issue:`28283`)
```
[the remark this answers is not in the document]
Comment (theOehrly) [reply doc/source/whatsnew/v1.4.2.rst:26]: Yes it is*, and sorry for the delayed response. I wanted to validate this again, more extensively. I've made some slight changes to the code, for testing purposes only, so that every single call to loc and iloc adds attrs to the original object and validates that they still exist on the returned object. After that, I ran the entire test suite with these additional checks built in. Assuming that this gives 100% coverage of all branches of loc and iloc, there were only three failures, all of which were related to _get_item_cache.
Basically, if attrs/_metadata/flags change after an object was cached, the next time a sliced object is returned using _get_item_cache, it will have the old attrs/_metadata/flags. I'm not sure if we'd want to add a finalize call to _get_item_cache though? If attrs/_metadata/flags are considered constants for each object, then this might not be a problem. This only came up because my extended testing modified attrs between multiple loc/iloc calls, but was never caused by a test itself.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3444,7 +3444,7 @@ def _ixs(self, i: int, axis: int = 0):
                 index=self.columns,
                 name=self.index[i],
                 dtype=new_values.dtype,
-            )
+            ).__finalize__(self)
```
[the remark this answers is not in the document]
Comment (theOehrly) [reply pandas/core/frame.py:3447]: Apart from the _iLocIndexer, this specific branch (axis=0) is only reached from FrameColumnApply.series_generator
There, the object that is returned by _ixs is only used internally to set its value (in FrameApply.apply_series_generator). It's therefore likely not necessary to call finalize for this?

We could call finalize in the _iLocIndexer. But that would be the only time that finalize is ever called in a _LocationIndexer. Which makes it seem like that's not the right place to do it. 

The else branch of _ixs always calls finalize in _box_col_values. Apparently this is the only place where _box_col_values is used.
We could therefore refactor _ixs so that internal calls that only set values on the returned object never call finalize in either branch. I'm not sure if there is a need for finalize to be called in these cases too, though? I guess not?

**Comment by theOehrly** [comment]
> This pull request is stale because it has been open for thirty days with no activity. Please [update](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#updating-your-pull-request) and respond to this comment if you're still interested in working on this.

Still interested in working on this, not stale.

@jreback how to proceed here?

**Comment by jreback** [comment]
you can merge master and update to any comments

**Comment by theOehrly** [comment]
> you can merge master and update to any comments

I was waiting for a response to my comments on the review above. I guess there still are open questions that should be addressed.

**Comment by jreback** [comment]
can look but pls merge master

**Comment by theOehrly** [comment]
@jreback main is merged and tests are fixed again

File: doc/source/whatsnew/v1.4.3.rst [hunk 3]
```diff
@@ -23,7 +23,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
--
+- Fixed metadata propagation in :meth:`NDFrame.xs`, :meth:`DataFrame.iterrows`, :meth:`DataFrame.loc` and :meth:`DataFrame.iloc` (:issue:`28283`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.3.rst:26]: pls move to 1.5

**Review by jreback — CHANGES_REQUESTED** [review]
It would be nice to run the performance benchmarks for indexing before/after for this change.

**Comment by theOehrly** [comment]
@jreback Results of the performance benchmark (only tests that got worse):

```
Benchmarks that have got worse:

       before           after         ratio
     [0141d921]       [f75a5dd8]
                      <finalize_indexers_iterators>
+      69.2±0.8μs       85.7±0.5μs     1.24  indexing.NonNumericSeriesIndexing.time_getitem_pos_slice('string', 'nonunique_monotonic_inc')
+       477±100μs         593±80μs     1.24  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int16Engine'>, <class 'numpy.int16'>), 'monotonic_decr', False, 100000)
+       462±100μs         635±40μs     1.38  indexing_engines.NumericEngineIndexing.time_get_loc_near_middle((<class 'pandas._libs.index.UInt8Engine'>, <class 'numpy.uint8'>), 'monotonic_decr', False, 100000)
```

Or the full output here: https://pastebin.com/raw/3sVMufVc

File: pandas/tests/generic/test_finalize.py [hunk 4]
```diff
@@ -259,10 +258,9 @@
     pytest.param(
         (
             pd.DataFrame,
-            frame_data,
-            operator.methodcaller("quantile", numeric_only=True),
+            ({"A": [pd.Timedelta(days=1), pd.Timedelta(days=2)]},),
```
Comment (jreback) [inline pandas/tests/generic/test_finalize.py:261]: why did this last change?

Comment (theOehrly) [reply pandas/tests/generic/test_finalize.py:261]: It was an exact duplicate of the test two tests up, and therefore redundant. I wanted to add this new test, so I just change this one.

**Comment by theOehrly** [comment]
@jreback any more feedback on this?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @theOehrly this is very nice!
